pkg/graph/crd/compat/schema.go

```diff
@@ -0,0 +1,220 @@
+	for propName, oldProp := range oldSchema.Properties {
+		propPath := path + ".properties." + propName
+
+		// check if property still exists
+		newProp, exists := newSchema.Properties[propName]
+		if !exists {
+			// property was removed - breaking change
+			result.AddBreakingChange(propPath, PropertyRemoved, "", "")
+			continue
+		}
+
+		// property exists in both schemas - compare them recursively
+		propResult, _ := compareSchemas(propPath, &oldProp, &newProp)
+		result.BreakingChanges = append(result.BreakingChanges, propResult.BreakingChanges...)
+		result.NonBreakingChanges = append(result.NonBreakingChanges, propResult.NonBreakingChanges...)
+	}
+
+	// Then check for added properties. Now things get a bit more spicy.
+	// A new property can be required or optional, and can have a default value.
+	// Depending on these factors, it can be a breaking or non-breaking change.
+	//
+	// In general the rules are:
+	// - Adding a required property without a default value is a breaking change
+	// - Adding a required property with a default value is a non-breaking change
+	// - Adding an optional property is a non-breaking change
+
+	newRequiredSet := toStringSet(newSchema.Required)
+
+	for propName, newProp := range newSchema.Properties {
+		if _, exists := oldSchema.Properties[propName]; !exists {
+			propPath := path + ".properties." + propName
+
+			// check if property is required but has a default (non-breaking)
+			// or required without default (breaking)
+			hasDefault := newProp.Default != nil && len(newProp.Default.Raw) > 0
+
+			if newRequiredSet[propName] && !hasDefault {
+				// property is required and has no default - breaking change
+				result.AddBreakingChange(propPath, PropertyAdded, "required=false", "required=true")
+			} else {
+				// property is optional or has default - non-breaking change
+				result.AddNonBreakingChange(propPath, PropertyAdded, "", "")
+			}
+		}
+	}
+}
+
+// compareRequiredFields checks for changes to required fields, it only considers
+// existing properties, since new properties are handled in compareProperties.
+func compareRequiredFields(path string, oldSchema, newSchema *v1.JSONSchemaProps, result *DiffResult) {
+	// Use length checks instead of nil checks
+	if len(oldSchema.Required) == 0 && len(newSchema.Required) == 0 {
+		return
+	}
+
+	// Convert to sets for efficient comparison
+	oldRequiredSet := toStringSet(oldSchema.Required)
+	newRequiredSet := toStringSet(newSchema.Required)
+
+	// Make a set of all existing property names (not newly added ones)
+	existingProps := make(map[string]bool)
+	for propName := range oldSchema.Properties {
+		existingProps[propName] = true
+	}
+
+	// Check for newly required fields ONLY for existing properties (breaking)
+	for req := range newRequiredSet {
+		// Only consider requirements for properties that already existed
+		if existingProps[req] && !oldRequiredSet[req] {
+			result.AddBreakingChange(path+".required", RequiredAdded, "", req)
```

Comment:
it is too bad this is not easier to use but Knative made `FieldError` to help with pathed errrors: https://github.com/knative/pkg/blob/b7bbf4be5dbd9a8b8b60e06094bd8930cf241357/apis/field_error.go#L63